app/src/main/java/com/example/development/sakaiclient20/ui/viewmodels/GradeViewModel.java

```diff
@@ -0,0 +1,113 @@
+package com.example.development.sakaiclient20.ui.viewmodels;
+
+import android.arch.lifecycle.LiveData;
+import android.arch.lifecycle.MutableLiveData;
+
+import com.example.development.sakaiclient20.persistence.entities.Grade;
+import com.example.development.sakaiclient20.repositories.CourseRepository;
+import com.example.development.sakaiclient20.repositories.GradesRepository;
+
+import java.util.HashMap;
+import java.util.List;
+
+import io.reactivex.android.schedulers.AndroidSchedulers;
+import io.reactivex.schedulers.Schedulers;
+
+public class GradeViewModel extends BaseViewModel {
+
+    private GradesRepository gradesRepository;
+    private HashMap<String, MutableLiveData<List<Grade>>> siteIdToGrades;
+
+    /**
+     * Grades view model constructor
+     *
+     * @param courseRepository course repository dependency needed for superclass
+     * @param gradesRepository grades repository dependency needed to refresh and get grades
+     */
+    public GradeViewModel(CourseRepository courseRepository, GradesRepository gradesRepository) {
+        super(courseRepository);
+        this.gradesRepository = gradesRepository;
+        this.siteIdToGrades = new HashMap<>();
+    }
+
+    /**
+     * Called by UI controller to get grades for a site
+     * if hashmap already has the grades, return that, otherwise
+     * refresh the grades and put into hashmap
+     *
+     * @param siteId site to get grades for
+     * @return live data containing grades list
+     */
+    public LiveData<List<Grade>> getGradesForSite(String siteId) {
+
+        if (!this.siteIdToGrades.containsKey(siteId)) {
+            this.siteIdToGrades.put(siteId, new MutableLiveData<>());
+            refreshSiteGrades(siteId);
+        }
+        return this.siteIdToGrades.get(siteId);
```

Comment:
why will returning an old livedata notify its observers about stale data? I thought the benefit of using livedata is that we can notify the observers when the data is changed, so wouldn't making a new one each time kind of defeat the purpose?

Reply:
When I was testing the courses stuff over the weekend, returning an existing `MutableLiveData` often triggered the observer's callback with the old data that was already in the `LiveData`. This happens because between returning the `LiveData` and `refreshData` 
completing, there is some time, and between that time the `LiveData` decides to trigger the callback since it already has some data that the observer is observing on.

You are definitely right about the purpose of `LiveData`, but I'm not sure why creating a new one defeats its purpose. We're simply abstracted the data layer from the UI, and the UI is listening for data that it needs (which in most of our cases is on a one-time basis, except for refreshing the main tabs), so I don't see how creating a new `LiveData` defeats its purpose, the UI is simply observing new data each time, but the _UI_ doesn't need to know about the details of the `ViewModel` and how it fetches data. All the UI cares about is observing on some data, receiving it, and modifying the UI.